beluga/include/beluga/algorithm/amcl_core.hpp

```diff
@@ -0,0 +1,240 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef AMCL_CORE_HPP
+#define AMCL_CORE_HPP
+
+#include <execution>
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <beluga/beluga.hpp>
+
+#include <range/v3/range/concepts.hpp>
+#include <range/v3/view/any_view.hpp>
+#include <range/v3/view/take_exactly.hpp>
+
+namespace beluga {
+
+/// Struct containing parameters for the Adaptive Monte Carlo Localization (AMCL) implementation.
+
+struct AmclParams {
+  /// Min distance in meters between updates.
+  double update_min_d = 0.25;
+  /// Min angular distance in radians between updates.
+  double update_min_a = 0.2;
+  /// Filter iterations interval at which a resampling will happen, unitless.
+  std::size_t resample_interval = 1UL;
+  /// Whether to use selective resampling or not.
+  bool selective_resampling = false;
+  /// Minimum number of particles in the filter at any point in time.
+  std::size_t min_particles = 500UL;
+  /// Maximum number of particles in the filter at any point in time.
+  std::size_t max_particles = 2000UL;
+  /// Used as part of the recovery mechanism when considering how many random particles to insert.
+  double alpha_slow = 0.001;
+  /// Used as part of the recovery mechanism when considering how many random particles to insert.
+  double alpha_fast = 0.1;
+  /// Used as part of the kld sampling mechanism.
+  double kld_epsilon = 0.05;
+  /// Used as part of the kld sampling mechanism.
+  double kld_z = 3.0;
+};
+
+/// Implementation of the Adaptive Monte Carlo Localization (AMCL) algorithm.
+/**
+ * \tparam MotionModel Class representing a motion model. Must satisfy \ref MotionModelPage.
+ * \tparam SensorModel Class representing a sensor model. Must satisfy \ref SensorModelPage.
+ * \tparam RandomStateGenerator A callable able to produce random states, optionally based on the current particles
+ * state.
+ * Class 'T' is a valid RandomStateGenerator if given 't' a possibly const instance of 'T' any of the following
+ * conditions are true:
+ * - t can be called without arguments returning an instance of 'SensorModel::state_type'
+ *   representing a random state.
+ * - t can be called with `const beluga::TupleVector<ParticleType>&> returning a callable
+ *   that can be called without arguments and return an instance of 'SensorModel::state_type'.
+ * \tparam WeightT Type to represent a weight of a particle.
+ * \tparam ParticleType Full particle type, containing state, weight and possibly
+ * other information .
+ * \tparam ExecutionPolicy Execution policy for particles processing.
+ */
+template <
+    class MotionModel,
+    class SensorModel,
```

Comment:
@serraramiro1 meta: as an afterthought, we could hide runtime type selection in perfect forwarding models. E.g. have a template motion model that takes a variant of motion models and forwards method and operator invocations using `std::visit()`. We would need some extra wrappers to deal with some lambda returns, but that's perfectly doable.

Then, this AMCL implementation would be the only one.